File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/client/renderer/particle/ParticleInstance.java
package pize.tests.voxelgame.client.renderer.particle;

import pize.graphics.texture.Region;
import pize.graphics.util.color.Color;
import pize.math.Maths;
import pize.math.vecmath.vector.Vec3f;
import pize.physic.Velocity3f;

public class ParticleInstance{
    
    private final Particle particle;
    private final long creationTime;
    private float elapsedTimeSeconds;
    private float alpha;
    
    public Region region;
    public float lifeTimeSeconds;
    public Vec3f position;
    public Velocity3f velocity;
    public float rotation, size;
    public Color color;
    
    public ParticleInstance(Particle particle, Vec3f position){
        this.region = new Region();
        this.particle = particle;
        this.position = position.copy();
        this.creationTime = System.currentTimeMillis();
        this.color = particle.color.copy();
        this.velocity = new Velocity3f();
    }
    
    public void update(){
        // Alpha
        elapsedTimeSeconds = (System.currentTimeMillis() - creationTime) / 1000F;
        alpha = particle.alphaFunc.apply(elapsedTimeSeconds / lifeTimeSeconds);
        alpha = Maths.clamp(alpha, 0, 1);
        // Animate
        particle.onAnimate.invoke(this);
    }
    
    public Particle getParticle(){
        return particle;
    }
    
    public float getElapsedTimeSeconds(){
        return elapsedTimeSeconds;
    }
    
    public float getAlpha(){
        return alpha;
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/client/renderer/particle/ParticleInstance.kt
package pize.tests.voxelgame.client.renderer.particle

import pize.graphics.texture.Region
import pize.graphics.util.color.Color
import pize.math.Maths.clamp
import pize.math.vecmath.vector.Vec3f
import pize.physic.Velocity3f

class ParticleInstance(val particle: Particle, position: Vec3f) {
    private val creationTime: Long
    var elapsedTimeSeconds = 0f
        private set
    var alpha = 0f
        private set
    var region: Region
    var lifeTimeSeconds = 0f
    var position: Vec3f
    var velocity: Velocity3f
    var rotation = 0f
    var size = 0f
    var color: Color

    init {
        region = Region()
        this.position = position.copy()
        creationTime = System.currentTimeMillis()
        color = particle.color.copy()
        velocity = Velocity3f()
    }

    fun update() {
        // Alpha
        elapsedTimeSeconds = (System.currentTimeMillis() - creationTime) / 1000f
        alpha = particle.alphaFunc.apply(elapsedTimeSeconds / lifeTimeSeconds)
        alpha = clamp(alpha, 0f, 1f)
        // Animate
        particle.onAnimate.invoke(this)
    }
}
